Flexible inline

facelessuser opened the pull request:
Add new InlineProcessor class that handles inline processing much better and allows for more flexibility.  This adds new InlineProcessors that no longer utilize unnecessary pretext and posttext captures.  New class can accept the buffer that is being worked on and manually process the text without regex and return new replacement bounds.  This helps us to handle links in a better way and handle nested brackets and logic that is too much for regular expression.  The refactor also allows image links to have links/paths with spaces like links. Ref #551, #613, #590, #161.

facelessuser committed:
Refactor inline patterns

Create a new inline Pattern2 class that uses a more effecient pattern.  Derive all inline classes from Pattern2, but leave the legacy Pattern, SimpleTextPattern, SimpleTagPattern, SubstituteTagPattern, and DoubleTagPattern for backwards compatibility.

facelessuser committed:
Rename new classes as InlineProcessor

Use the more appropriate `InlineProcessor` name for new classes opposed to the legacy `Pattern2`.

facelessuser committed:
Increase flexibility of InlineProcessor

Allow additional analysis by passing in data buffer into handleMatch.  This allows us to give a simple regex for a start token, and then we can tokenize the content if we like inside handleMatch.  If we don't find what we are looking for, we can reject the match as if we found nothing. You must return your element and the bounds of the handled characters.

facelessuser committed:
Faster string iterating

facelessuser committed:
Fix indentation

facelessuser committed:
Python2 finditer doesn't take keyword args

facelessuser committed:
Handling of recursive round brackets in links

Initial handling of recursive round brackets in link's href.

facelessuser committed:
Fix variable type and add test for #613

facelessuser committed:
Allow spaces in image URL and cleanup

Clean up some variable naming, and perfrom comon actions on returned text before returning it.  Also remove code that turns off spaces in URL to allow images to behave like links with spaces in its URL.

facelessuser committed:
Rename inline processors

For consistency, rename Pattern classes to InlineProcessor.

facelessuser committed:
Tweaks to url handling with titles

Handle titles in a more sane way

facelessuser committed:
Update docstrings and cleanup

Remove unused variable and update docstring describing functionality in better detail.

facelessuser committed:
Merge branch 'master' into flexible-inline

facelessuser committed:
Use new unittest framework

facelessuser committed:
Remove old variable and comments

facelessuser committed:
Merge branch 'master' into flexible-inline

facelessuser committed:
Remove sanitize_url

facelessuser committed:
Merge branch 'master' into flexible-inline

facelessuser committed:
Add single quote test case

facelessuser committed:
Don't add link test to legacy files

facelessuser commented:
Figured I'd get some discussion going on this and start moving this forward.

facelessuser committed:
Update html of legacy with removed lines

facelessuser commented:
Seems pypy is failing for the recently discussed issue Future warnings with elementtree:

```
FutureWarning: The behavior of this method will change in future versions.  Use specific 'len(elem)' or 'elem is not None' test instead.
```

While this is a separate issue, I might take care of it here as well.

facelessuser committed:
Fix elementtree Future warning

facelessuser committed:
Remove dead patterns

waylan commented:
> Seems pypy is failing for the recently discussed issue Future warnings with elementtree:

Yeah, we keep fixing these, and then every time we do more work on ElementTree related code, we introduce them. So I made sure a test fails to stop us from reintroducing the warnings. I consider these warnings a reason to not accept a PR, so yeah, they should be addressed (which I see you did).

waylan commented on tests/misc/image.html:
```diff
@@ -1,5 +1,5 @@
-<p>![Fail](http://humane man.jpg)</p>
\ No newline at end of file
+<p><img alt="Image" src="http://humane man.jpg" title="The most humane man." /></p>
+<p><img alt="Image" src="http://humane man.jpg" /></p>
```
Rather than editing these, lets delete them and recreate the tests in the new syntax tests.

facelessuser replied:
Makes sense. I can probably move the whole file to the new format.

waylan replied:
The whole file woull be even better

waylan commented on tests/misc/image.txt:
```diff
@@ -7,6 +7,6 @@
+![Image](http://humane man.jpg "The most humane man.")
 
-![Fail](http://humane man.jpg)
+![Image](http://humane man.jpg)
```
Rather than editing these, lets delete them and recreate the tests in the new syntax tests.

waylan commented on tests/test_syntax/inline/test_advanced_images.py:
```diff
@@ -0,0 +1,99 @@
+from markdown.test_tools import TestCase
+
+
+class TestAdvancedImages(TestCase):
```
I suppose we can leave the class named "advanced" but let's just name the file `images.py`. Eventually all "image" related tests should be moved here.

facelessuser replied:
Sounds good.

waylan commented on tests/test_syntax/inline/test_advanced_links.py:
```diff
@@ -0,0 +1,98 @@
+from markdown.test_tools import TestCase
+
+
+class TestAdvancedLinks(TestCase):
```
I suppose we can leave the class named "advanced" but let's just name the file `links.py`. Eventually all "link" related tests should be moved here.

facelessuser replied:
Sounds good.

waylan reviewed (changes requested):
I still need to take a closer look at what some of the code is actually doing, but here are a few general organization issues I spotted. Overall it looks good. Thanks for doing the work on this.

waylan commented:
Oh and while it may seem obvious, I should note that the docs need to be updated for these changes. Specially, the Extension API docs.

facelessuser commented:
Makes sense. It's a pretty big change. I figured it'd take a little bit to work through the review.

facelessuser commented:
Yeah, I didn't make changes to the docs yet only because I wanted to make sure we were all in agreement on the new processors behavior. Figured I'd wait until we were good with the code.

facelessuser committed:
Merge legacy misc test with new test

Merge legacy test misc/image.txt with new (renamed) inline/images.py.

facelessuser committed:
API docs updated with new processor info

Describe the new InlineProcessor and its usage in the documents.

facelessuser commented:
Went ahead and addressed the test changes suggested, and provided some API documentation.  I'll address spelling issues if Travis notifies of any.

facelessuser committed:
Fix lint and spelling failures

waylan commented on docs/extensions/api.md:
```diff
@@ -110,8 +112,113 @@ implemented with separate instances of the `SimpleTagPattern` listed below.
+    start with `r'^(.*?)'` and end with `r'(.*?)!'`. This was a huge
+    performance sink and this requirement has been removed. The returned match
+    object will only contain what is explicitly matched in the pattern, and
+    extension pattern groups now starts with `m.group(1)`.
```
should be ``extension pattern groups now start with `m.group(1)`.``

You have an extra `s` on the end of `start`.

facelessuser replied:
Cool, will fix.

waylan commented on docs/extensions/api.md:
```diff
@@ -110,8 +112,113 @@ implemented with separate instances of the `SimpleTagPattern` listed below.
+
+    Returns an element of type "`tag`" with a text attribute of `group(3)`
+    of a `pattern`. `tag` should be a string of a HTML element (i.e.: 'em').
+    Also returns the boundaries of where the data to replace resides.
```
> returns the boundaries of where the data to replace resides.

I understand that last line, but if feels very awkward. I had to re-read it a couple times the first time I saw it. Same applies for the other few places it is used. Sorry, I don't have a better suggestion.

facelessuser replied:
Yeah, I'll rephrase it.

waylan reviewed (changes requested):
Thanks for the documentation. I noted a couple minor issues. Otherwise looks good

facelessuser committed:
Documentation revisions

Fix unnecessary plural usage of start, and rephrase processor return sentences.

waylan commented:
@facelessuser, if you rebase this against master (resolving merge conflicts), I think we can accept this.

facelessuser commented:
Cool, I'll pull the latest changes into this pull request and resolve any conflicts.  I've pulled in most of the changes already, so imagine it should go fairly smooth.

facelessuser committed:
Merge branch 'master' into flexible-inline

facelessuser committed:
Skip coverage on unused legacy code for now

facelessuser commented:
Branch is synced up.
